community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java

```diff
@@ -378,35 +382,137 @@ public void createIndex( IndexRule rule )
+    {
+        boolean exceptionThrown = false;
+        try
+        {
+            IndexUpdater updater = updaterMap.getUpdater( descriptor );
+            Reservation reservation = updater.validate( indexUpdates );
+            aggregatedReservation.add( reservation );
+        }
+        catch ( IOException | IndexCapacityExceededException e )
+        {
+            exceptionThrown = true;
+            String indexName = descriptor.userDescription( tokenNameLookup );
+            throw new UnderlyingStorageException( "Validation of updates for index " + indexName + " failed", e );
+        }
+        catch ( Throwable t )
+        {
+            exceptionThrown = true;
+            throw t;
+        }
+        finally
+        {
+            if ( exceptionThrown )
             {
-                // This is a temporary measure to resolve a corruption bug. We believe that it's caused by stray
-                // HA transactions, and we know that this measure will fix it. It appears, however, that the correct
-                // fix will be, as it is for several other issues, to modify the system to allow us to kill running
-                // transactions before state switches.
-                throw new IllegalStateException( "Cannot queue index updates while index service is " + state );
+                aggregatedReservation.release();
             }
         }
     }
 
-    protected void applyRecoveredUpdates() throws IOException
+    private static ValidatedIndexUpdates newValidatedIndexUpdates( final IndexUpdaterMap indexUpdaters,
+            final Map<IndexDescriptor,List<NodePropertyUpdate>> updatesByIndex, final Reservation reservation )
+    {
+        return new ValidatedIndexUpdates()
+        {
+            @Override
+            public void flush() throws IOException, IndexEntryConflictException, IndexCapacityExceededException
+            {
+                for ( Map.Entry<IndexDescriptor,List<NodePropertyUpdate>> entry : updatesByIndex.entrySet() )
+                {
+                    IndexDescriptor indexDescriptor = entry.getKey();
+                    List<NodePropertyUpdate> updates = entry.getValue();
+
+                    IndexUpdater updater = indexUpdaters.getUpdater( indexDescriptor );
+                    for ( NodePropertyUpdate update : updates )
+                    {
+                        updater.process( update );
+                    }
+                }
+            }
+
+            @Override
+            public void close()
+            {
+                try
+                {
+                    reservation.release();
+                }
+                finally
+                {
+                    indexUpdaters.close();
+                }
+            }
+        };
+    }
+
+    private void applyRecoveredUpdates() throws IOException
     {
         logger.debug( "Applying recovered updates: " + recoveredNodeIds );
```

Comment:
Might want to guard this one with `isDebugEnabled` since the set might be large.